### PR TITLE
feat: Add a CLI executable for code generation 

### PR DESCRIPTION
**Pull request by calvincestari**
PR 3 of #2166 

This PR:
* adds the command-line executable as a new Swift package project in the `./CodegenCLI` directory
* adds a test step to CI to verify the executable builds on each change. The test output for this PR can be found [here](https://app.circleci.com/pipelines/github/apollographql/apollo-ios/3156/workflows/42ec5487-6b64-428a-8976-9a1c35aba46b/jobs/26126/parallel-runs/0/steps/0-105).

Next PR:
* make it available on Homebrew
* refine Release build settings

Below is some sample output of the commands and parameters:
```
❯ .build/debug/apollo-ios-codegen --help
OVERVIEW: A command line utility for Apollo iOS code generation.

USAGE: apollo-ios-codegen <subcommand>

OPTIONS:
  --version               Show the version.
  -h, --help              Show help information.

SUBCOMMANDS:
  init                    Initialize a new configuration with defaults.
  validate                Validate a configuration file or JSON formatted string.
  generate                Generate Swift source code based on a code generation configuration.

  See 'apollo-ios-codegen help <subcommand>' for detailed help.
```
```
❯ .build/debug/apollo-ios-codegen init --help
OVERVIEW: Initialize a new configuration with defaults.

USAGE: apollo-ios-codegen init [--output <output>] [--path <path>] [--overwrite]

OPTIONS:
  -o, --output <output>   Destination for the new configuration. (default: file)
  -p, --path <path>       Write the configuration to a file at the path. (default: ./apollo-codegen-config.json)
  -w, --overwrite         Overwrite any file at --path.
  --version               Show the version.
  -h, --help              Show help information.
```
```
❯ .build/debug/apollo-ios-codegen validate --help
OVERVIEW: Validate a configuration file or JSON formatted string.

USAGE: apollo-ios-codegen validate --input <input> [--path <path>] [--string <string>]

OPTIONS:
  -i, --input <input>     Configuration source.
  -p, --path <path>       Read the configuration from a file at the path.
  -s, --string <string>   Configuration string in JSON format.
  --version               Show the version.
  -h, --help              Show help information.
```
```
❯ .build/debug/apollo-ios-codegen generate --help
OVERVIEW: Generate Swift source code based on a code generation configuration.

USAGE: apollo-ios-codegen generate [--input <input>] [--path <path>] [--string <string>]

OPTIONS:
  -i, --input <input>     Configuration source. (default: file)
  -p, --path <path>       Read the configuration from a file at the path. (default: ./apollo-codegen-config.json)
  -s, --string <string>   Configuration string in JSON format.
  --version               Show the version.
  -h, --help              Show help information.
```

### CODE REVIEW

**Comment by martinbonnin** [comment]
This is so cool! Is it possible to download it standalone by any chance ? Something like `brew install apollo-ios-codegen`?

Also might be worth dropping the `-codegen` suffix if the CLI supports other commands later on? (`downloadSchema`, `transformOperations`, ...)

**Comment by calvincestari** [comment]
Thanks @martinbonnin! Yes, the next task in this issue is to make it available via Homebrew.

I'm cautious about extending the functionality to things that might cross-over with [Rover](https://www.apollographql.com/docs/rover/graphs/); it seems purpose built for those kinds of tasks. For now it's focussed on Swift code generation.

**Comment by AnthonyMDev** [comment]
I agree that we shouldn't cross-over too much with Rover, but there might be some value in an option to download the schema. Perhaps we can actually use Rover for that internally? We already have `ApolloSchemaDownloader`. Unless we are going to get rid of that, it should probably be part of the CLI.

@martinbonnin What would a `transformOperations` CLI command do exactly? Is this something you have in kotlin?

File: CodegenCLI/.gitignore [hunk 1]
```diff
@@ -0,0 +1,9 @@
+.DS_Store
```
Comment (AnthonyMDev) [inline CodegenCLI/.gitignore:1]: Does the CLI directory need it's own gitignore file? I thought gitignore only worked in the repo root?

File: CodegenCLI/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
+# CodegenCLI
```
Comment (AnthonyMDev) [inline CodegenCLI/README.md:1]: Awesome that you created this Readme!

File: CodegenCLI/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+```
+
+## Generate
+```
+OVERVIEW: Generate Swift source code based on a code generation configuration.
+
+USAGE: apollo-ios-codegen generate [--input <input>] [--path <path>] [--string <string>]
+
+OPTIONS:
+  -i, --input <input>     Configuration source. (default: file)
```
Comment (AnthonyMDev) [inline CodegenCLI/README.md:55]: Same as above, I'm feeling like it's redundant to have an input option if it also requires you to then provide either `--path` or `--string`.

File: CodegenCLI/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+## Initialize
+```
+OVERVIEW: Initialize a new configuration with defaults.
+
+USAGE: apollo-ios-codegen init [--output <output>] [--path <path>] [--overwrite]
+
+OPTIONS:
+  -o, --output <output>   Destination for the new configuration. (default: file)
+  -p, --path <path>       Write the configuration to a file at the path. (default: ./apollo-codegen-config.json)
+  -w, --overwrite         Overwrite any file at --path.
```
Comment (AnthonyMDev) [inline CodegenCLI/README.md:29]: Can we add "If init is called without --overwrite and a config file already exists at the output path, the command will fail."

File: CodegenCLI/README.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+```
+
+## Initialize
+```
+OVERVIEW: Initialize a new configuration with defaults.
+
+USAGE: apollo-ios-codegen init [--output <output>] [--path <path>] [--overwrite]
+
+OPTIONS:
+  -o, --output <output>   Destination for the new configuration. (default: file)
```
Comment (AnthonyMDev) [inline CodegenCLI/README.md:27]: Do we need this? If they provide a path, we can assume they want to write it to a file at that path. If they don't assume we are printing it to stdout.

Maybe in that case, change the `--path` option to be the `--output` option and get rid of the existing output option?

**Review by AnthonyMDev — COMMENTED** [review]
The code looks super clean and love the documentation. Just some questions about the shape of the cli command options. We should also consider including the schema downloader!

File: CodegenCLI/.gitignore (hunk 1 above)
Comment (calvincestari) [reply CodegenCLI/.gitignore:1]: You're correct, it doesn't. I just did a quick test and it's actually created by the `swift package init` command. My mistake for not being more aware of it getting committed - I'll get rid of it.

**Comment by martinbonnin** [comment]
> What would a transformOperations CLI command do exactly? Is this something you have in kotlin?

We don't have any CLI in Kotlin but I was thinking it could be useful for codegen and maybe other things. `transformOperations` would be used for safelisting to expose the queries after all transformations (__typename, key fields, etc...).

File: .swiftpm/xcode/xcshareddata/xcschemes/Apollo.xcscheme [hunk 6]
```diff
@@ -1,113 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (calvincestari) [inline .swiftpm/xcode/xcshareddata/xcschemes/Apollo.xcscheme:1]: It's odd that these are deleted but note that this is the `.swiftpm` directory. I checked the apollo-ios Xcode project, in this branch, and the schemes are still there and shared so it must just be an Xcode SPM project thing.

**Pull request merged**